flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -1697,133 +1621,620 @@ class ExprLowering {
+      return builder.create<OP>(getLoc(), left, right);
+    };
+  }
+
+#undef GENBIN
+#define GENBIN(GenBinEvOp, GenBinTyCat, GenBinFirOp)                           \
+  template <int KIND>                                                          \
+  CC genarr(const Fortran::evaluate::GenBinEvOp<Fortran::evaluate::Type<       \
+                Fortran::common::TypeCategory::GenBinTyCat, KIND>> &x) {       \
+    return createBinaryOp<GenBinFirOp>(x);                                     \
   }
 
+  GENBIN(Add, Integer, mlir::AddIOp)
+  GENBIN(Add, Real, fir::AddfOp)
+  GENBIN(Add, Complex, fir::AddcOp)
+  GENBIN(Subtract, Integer, mlir::SubIOp)
+  GENBIN(Subtract, Real, fir::SubfOp)
+  GENBIN(Subtract, Complex, fir::SubcOp)
+  GENBIN(Multiply, Integer, mlir::MulIOp)
+  GENBIN(Multiply, Real, fir::MulfOp)
+  GENBIN(Multiply, Complex, fir::MulcOp)
+  GENBIN(Divide, Integer, mlir::SignedDivIOp)
+  GENBIN(Divide, Real, fir::DivfOp)
+  GENBIN(Divide, Complex, fir::DivcOp)
+
+  template <Fortran::common::TypeCategory TC, int KIND>
+  CC genarr(
+      const Fortran::evaluate::Power<Fortran::evaluate::Type<TC, KIND>> &x) {
+    return [](IterSpace iters) -> ExtValue {
+      return mlir::Value{}; /* FIXME */
+    };
+  }
+  template <Fortran::common::TypeCategory TC, int KIND>
+  CC genarr(
+      const Fortran::evaluate::Extremum<Fortran::evaluate::Type<TC, KIND>> &x) {
+    return [](IterSpace iters) -> ExtValue {
+      return mlir::Value{}; /* FIXME */
+    };
+  }
+  template <Fortran::common::TypeCategory TC, int KIND>
+  CC genarr(
+      const Fortran::evaluate::RealToIntPower<Fortran::evaluate::Type<TC, KIND>>
+          &x) {
+    return [](IterSpace iters) -> ExtValue {
+      return mlir::Value{}; /* FIXME */
+    };
+  }
+  template <int KIND>
+  CC genarr(const Fortran::evaluate::ComplexConstructor<KIND> &x) {
+    return [](IterSpace iters) -> ExtValue {
+      return mlir::Value{}; /* FIXME */
+    };
+  }
+  template <int KIND>
+  CC genarr(const Fortran::evaluate::Concat<KIND> &x) {
+    TODO("concat");
+    return [](IterSpace iters) -> ExtValue { return mlir::Value{}; };
+  }
+  template <int KIND>
+  CC genarr(const Fortran::evaluate::SetLength<KIND> &x) {
+    TODO("set length");
+    return [](IterSpace iters) -> ExtValue { return mlir::Value{}; };
+  }
   template <typename A>
-  void body(const Fortran::evaluate::Expr<A> &exp) {
-    (void)builder;
-    (void)symMap;
-    (void)loopCounters;
-    (void)arrayBlockArgs;
-    (void)loopReturnVals;
-    (void)exprCtx;
-    TODO("");
+  CC genarr(const Fortran::evaluate::Constant<A> &x) {
+    return [](IterSpace iters) -> ExtValue {
+      return mlir::Value{}; /* FIXME */
+    };
+  }
+  CC genarr(const Fortran::evaluate::ArrayRef &arr) {
+    return [](IterSpace iters) -> ExtValue {
+      TODO("array ref");
+      return mlir::Value{};
+    };
+  }
+  CC genarr(const Fortran::semantics::SymbolRef &sym) {
+    auto extMemref = symMap.lookupSymbol(sym).toExtendedValue();
```

Comment:
Should this call `gen(const Fortran::semantics::SymbolRef &sym)` to get the ExtendedValue for the symbols ? I think the logic should be the same whether in array context or not, and I may have to add a few allocatable/pointer related things there. I can of course duplicate the logic if you prefer to keep things separated, it's just a few things.